### ISSUE TITLE
How to implement a platform-agnostic gpio/i2c/spi wasi api?

### ISSUE THREAD

**Issue by rahul-thakoor** [WebAssembly/WASI#443]
Hello!

I am very new to the wasm/wasi ecosystem, but I'm interested in learning some more and help with such an implementation if possible. I saw there is a related closed issue #251 

Context:
I have a similar use case where I could like to be able to access GPIO pins or even i2c and SPI based devices from a wasm module. I think using the existing wasi filesystem api I can achieve something like this to some extent. E.g I can implement the linux sysfs-gpio interface partially but then the wasm module is not portable per se - it would only run on linux host for instance.

I am trying to understand how to make such a platform-agnostic API such that the same module works on Linux, Windows, macOS, or even microcontroller hosts as long as there is a runtime that implements the hardware abstraction API.

It feels like there is indeed some overlap with the rust-embedded [embedded-hal library](https://crates.io/crates/embedded-hal) and the API could be inspired by that.

Not sure if this is the right place for such a discussion. Feel free to redirect accordingly 😃

Thanks 👍

**Comment by sunfishcode** [comment]
Something like embedded-hal indeed sounds like a good place to start. In general, WASI prefers to use typed APIs rather than filesystem-oriented abstractions, in part because of the portability issues you mention.

The current official way to write WASI APIs is to write an s-expression-based witx interface, and there are a few tools for using that to generate bindings for various languages. That said, these tools aren't very mature, and at this point are somewhat out of date with respect to interface types' evolution. I suggest taking a look at [witx-bindgen](https://github.com/bytecodealliance/witx-bindgen), which is a new tool still in development, but it learns several lessons from earlier witx work, and incorporates new ideas like the canonical ABI. Please feel free to ask questions!

**Comment by rahul-thakoor** [comment]
Thanks for the clarification @sunfishcode 

So if I understand correctly, say I were to implement the embedded-hal traits for wasi, I would
1. Need to describe the API in `.witx` format
2. Use `witx-bindgen` to generate the bindings
3. Implement the bindings for the runtime I want to use

Is that correct?

I'm not 100% sure how to implement something like gpio access using typed APIs and also make that cross-platform. Is it a case of implementing the traits generated separately for each target platform? is there a reference gpio or other library that works for both unix and windows which can be used as a reference?

Thanks

**Comment by sunfishcode** [comment]
Yes, that sequence sounds correct. As I mentioned above, the tools are still in development, but that seems like the best path at this time, and as it matures it'll get better.

The nature of WASI is that we'll typically want to use the same interface for all platforms. Can you say more about what makes gpio difficult do in a cross-platform way?

**Comment by rahul-thakoor** [comment]
> Can you say more about what makes gpio difficult do in a cross-platform way?

I'm not sure, to be honest. But I'm thinking for implementing the traits for each platform, we will have to use platform specific apis which do provide different interfaces imo.

> The nature of WASI is that we'll typically want to use the same interface for all platforms. 

Is there an example of perhaps something simple that uses typed APIs to blink a system led for instance? is this possible by using existing wasi api only?

**Comment by rahul-thakoor** [comment]
Was trying to look for some more info on canonical ABI? is this https://github.com/WebAssembly/interface-types/blob/40f157ad429772c2b6a8b66ce7b4df01e83ae76d/proposals/interface-types/CanonicalABI.md the most up to date one please?

**Comment by SamuraiCrow** [comment]
Would a volatile address space be the best way to implement a sandboxed memory-mapped I/O be the best way to implement simple devices to a static driver?  I see there's a proposal for multiple address spaces on https://github.com/WebAssembly/multi-memory which might be a means to the end we are seeking.

**Comment by emielvanseveren** [comment]
What steps should I take to get things started around this api? Should I plan something in the biweekly wasi meetings or create an initial wit impl first?

**Comment by shaggygi** [comment]
Might want to keep an eye on dotnet\iot as they will most likely add gpio controller and i2c devices for this.  I'd say it would be a ways off until the core wasi stuff matures.

**Comment by linclark** [comment]
@emielvanseveren you can find more information on how to champion a proposal in the [Contributing guide](https://github.com/WebAssembly/WASI/blob/main/Contributing.md)

**Comment by joakimeriksson** [comment]
Anyone started to work on this yet? (we are looking into support WASM on IoT-devices). I guess one question is on what level do we want the API's to be at? GPIO/SPI means that device drivers will be in WASM also - but if we go for sensors the drivers will be "below" / provided by host. On the other hand - we could have both to provide WASI-defined APIs from within a WASM instance...

**Comment by emielvanseveren** [comment]
@joakimeriksson I have planned an agenda item on the 9th of March to approve this for phase 1. That is a good question and is something that needs to be discussed further. My initial thought would be to support both although I'm not sure if moving the drivers to the host fits with the platform agnostic goal of WASI.

**Comment by emielvanseveren** [comment]
> 
You mentioned in an earlier reply that MMIO might be achievable with multi-memory. Any further thoughts regarding that aspect?

**Comment by SamuraiCrow** [comment]
@emielvanseveren The issue there, https://github.com/WebAssembly/multi-memory/issues/9 , is closed but discussed in a different proposal:  https://github.com/WebAssembly/memory-control .

**Comment by sunfishcode** [comment]
Another option for MMIO is to have host functions which do the actual loads and stores. This has several nice properties: it abstracts away the need for the addresses to live inside the linear memory address range, it's virtualizable so you can write a hardware emulator just by implementing the functions rather than having to hook into the Wasm engine to intercept loads and stores, it doesn't require us adding I/O semantics to Wasm linear memory, and the host code can be in full control of the sizes and alignments of the accesses to the MMIO memory.

**Comment by joakimeriksson** [comment]
@sunfishcode - I would go for this type of abstraction - it would give a more high-level interface to the applications also and still enable "higher-level" drivers for some types of hardware (if done right). We are considering adding WASM and WASI in an IoT-OS and are looking into how the abstractions should be in this context.

**Comment by woodsmc** [comment]
To help guide us in making decisions on the level of abstraction I think considering the min footprint of an embedded device and some specific use cases we want to support first will help. Perhaps also sharing some of this thinking with the WAMR TSC would be a great way to extend the level of feedback. 

The approach @sunfishcode suggests, is what I was mentioning on the Meeting on March 9th. You see the same approach in Microkenel OS architectures. A separation of logic, and physical device drivers, and then these also wrapped and then exposed as services (known interfaces) to the application code (wasm). So something like this... (I'm making this up as an illustration for a disk and file system)...

* Physical layer: Software handling Interrupt handling, DMA systems, GPIO's etc. (Native / host system)
* Logical Layer: Software implementing Block Storage Drive Controller (given the physical controller for any number of drives can provide block storage) (Native / host system)
* Interface Layer: File System (fopen / fclose etc. behind the scenes it's implementing a exFAT format) (Native / host system)
 *Application Layer: WASM.

This works, but has an overhead which will impose minimum requirements on the platforms that can support WASM. The stratification above is often rejected by RTOS implementations who provide a bare-metal `fopen` function so that they can support as many smaller embedded systems as possible.

Providing a definition of what we mean by an embedded system will help us determine the correct layer of abstraction we'd like to provide. Some of this can be determined by the existing minimum system requirements that, say WAMR has and some of the use cases that the WAMR TSC may want to support.

**Comment by yamt** [comment]
is this about something at the same layer as https://docs.oracle.com/javame/8.2/api/dio/api/index.html ?

**Comment by joakimeriksson** [comment]
Is there any activities around this? We are putting some efforts into experimenting with WASM in Contiki-NG - for small/resource constrained IoT devices. It would be great having a discussion on this topic and do a reference implementation - we would be happy to discuss, evaluate and test!

**Comment by sunfishcode** [comment]
I put together a rough sketch of a gpio/i2c/spi API, and built a prototype:
 - An [API description](https://github.com/sunfishcode/hello-embedded/blob/main/wit/embedded.wit), which includes GPIO, I²C, PWM, SPI, and timers.
 - A  [repo with a simple example guest, and a host simulator that can run it](https://github.com/sunfishcode/hello-embedded).

**Comment by Zelzahn** [comment]
Greetings everyone, I have submitted a PR to update the I2C proposal ([PR](https://github.com/WebAssembly/wasi-i2c/pull/1)). Currently, the WIT is largely based upon the works of @sunfishcode.

**Comment by Enet4** [comment]
I just stumbled upon this while looking at the [list of feature proposals in stage 1](https://wasi.dev/interfaces#phase-1---feature-proposal-cg). The repository [wasi-spi](https://github.com/WebAssembly/wasi-spi) was created for this proposal, but it's still in an initial template-backed state. Could some of the information from this issue be transmitted to that repo, even if just to clarify the intent behind this proposal? The acronym SPI without any context is a bit ambiguous.

**Comment by SamuraiCrow** [comment]
@Enet4 I think SPI stands for serial peripheral interface. It's used by embedded controllers like the Arduino.

**Comment by Enet4** [comment]
> I think SPI stands for serial peripheral interface. It's used by embedded controllers like the Arduino.

Yeah, my point was that if someone goes through the [WASI proposals page](https://wasi.dev/interfaces#phase-1---feature-proposal-cg), they need to go out of their way to search for the [source of this proposal](https://github.com/WebAssembly/meetings/blob/61ec6f1689a47d7eac99309a12fca38fd5826b9e/wasi/2023/WASI-03-09.md?plain=1#L34) and find this issue. Before that, I had no idea. There are still no direct references from `wasi-spi`, and there are technical contexts where SPI would stand for something else.

I do not have a stake on this, but at the current state of affairs, one cannot even know whether the proposal is relevant from inspecting the repository, hence my suggestion to transfer some information here. Maybe I can assist in this, but I believe someone more familiar with the proposal would do a better job at it. 😅 I filed a starting point for now: https://github.com/WebAssembly/wasi-spi/issues/1

**Comment by ricochet** [comment]
The path, for anyone reading the thread later:

- @sunfishcode sketched a combined [`embedded.wit`](https://github.com/sunfishcode/hello-embedded/blob/main/wit/embedded.wit) covering GPIO, I²C, PWM, SPI and timers, with a working guest and host simulator
- @Zelzahn turned the I²C portion into [wasi-i2c#1](https://github.com/WebAssembly/wasi-i2c/pull/1), based largely on that sketch
- Those became separate proposals

Where they stand today:

| Proposal | Phase | Champions | WIT |
|---|---|---|---|
| [wasi-i2c](https://github.com/WebAssembly/wasi-i2c) | 2 | Friedrich Vandenberghe, Merlijn Sebrechts, Maximilian Seidler | `wasi:i2c@0.2.0-draft` |
| [wasi-gpio](https://github.com/WebAssembly/wasi-gpio) | 1 | Merlijn Sebrechts, Maximilian Seidler | `wasi:gpio@0.2.0`, with `general`, `digital` and `analog` interfaces |
| [wasi-spi](https://github.com/WebAssembly/wasi-spi) | 1 | Merlijn Sebrechts | `wasi:spi` |

@Enet4 your point from last February is resolved: `wasi-spi` is no longer template-backed, it now has a real [`spi.wit`](https://github.com/WebAssembly/wasi-spi/blob/main/wit/spi.wit).

Two parts of the original sketch still have no home: **PWM** and **timers**. There is no `wasi-pwm` proposal. If either matters to you, that is worth opening as a new pre-proposal, see [CONTRIBUTING.md](https://github.com/WebAssembly/WASI/blob/main/CONTRIBUTING.md#the-phase-process).

Design discussion should now go to the individual repos, where it reaches the champions directly. Reopening is always welcome if something here was missed.
